x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Subject.java

```diff
@@ -231,6 +232,25 @@ private RoleReferenceIntersection buildRoleReferencesForApiKey() {
         );
     }
 
+    private RoleReferenceIntersection buildRoleReferencesForRemoteAccess() {
+        final List<RoleReference> roleReferences = new ArrayList<>(buildRoleReferencesForApiKey().getRoleReferences());
+        @SuppressWarnings("unchecked")
+        final List<RemoteAccessAuthentication.RoleDescriptorsBytes> remoteAccessRoleDescriptorsBytes = (List<
+            RemoteAccessAuthentication.RoleDescriptorsBytes>) metadata.get(AuthenticationField.REMOTE_ACCESS_ROLE_DESCRIPTORS_KEY);
+        if (remoteAccessRoleDescriptorsBytes.isEmpty()) {
+            // If the remote access role descriptors are empty, the remote user has no privileges. We need to add an empty role to restrict
+            // access of the overall intersection accordingly
+            roleReferences.add(new RoleReference.RemoteAccessRoleReference(RemoteAccessAuthentication.RoleDescriptorsBytes.EMPTY));
+        } else {
+            // TODO handle this once we support API keys as querying subjects
+            assert remoteAccessRoleDescriptorsBytes.size() == 1;
+            for (RemoteAccessAuthentication.RoleDescriptorsBytes roleDescriptorsBytes : remoteAccessRoleDescriptorsBytes) {
+                roleReferences.add(new RoleReference.RemoteAccessRoleReference(roleDescriptorsBytes));
```

Comment:
It is worthwhile to insert the remote roleReferences in front of the API key's roleReferences. When the role is built by `RoleReferenceIntersection.buildRole`, it first builds a role from the first roleReference from the list. The role is then limited by the role built from the second roleReference and so on so forth. Though the authorization result does not change if the order is reversed, conceptually I think it is still useful to have the "limited by" model. 

In this case, it means the remote roleReference is limited by the API key's assigned roleReference, then further limited by the API key's limited-by roleReference. If the remote is also an API key, it will then be "[remote_assigned_role, remote_limited_by_role, apikey_assigned_role, apikey_limited_by_role]".

What do you think?

Reply:
I'm good with flipping the order -- in my mind, it was the remote access API key being limited by its owner user's privileges and then further limited by the remote user's privileges, but the inverse also makes sense 👍